sw/airborne/modules/ctrl/optical_flow_hover.c

```diff
@@ -0,0 +1,736 @@
+
+  covariances.Z = 0.0f;
+  cov_array_filledZ = 0;
+
+  of_hover_ctrl_Z.PID.P = OFH_PGAINZ;
+
+  of_hover_ctrl_Z.errors.sum_err = 0.0f;
+  of_hover_ctrl_Z.errors.d_err = 0.0f;
+  of_hover_ctrl_Z.errors.previous_err = 0.0f;
+
+  vision_time = get_sys_time_float();
+  prev_vision_timeZ = vision_time;
+}
+
+// Read H RC
+void guidance_h_module_read_rc(void) {}
+
+/**
+ * Run the horizontal optical flow hover module
+ */
+void horizontal_ctrl_module_run(bool in_flight)
+{
+  /***********
+   * TIME
+   ***********/
+  float ventral_factor = -1.28f; // magic number comprising field of view etc.
+
+  float dt = vision_time - prev_vision_timeXY;
+
+  // check if new measurement received
+  if (dt <= 1e-5f) {
+    return;
+  }
+
+  /***********
+   * VISION
+   ***********/
+
+  float lp_factor = dt / OF_LP_CONST;
+  Bound(lp_factor, 0.f, 1.f);
+
+  float new_flowX = (flowX * ventral_factor) / dt;
+  float new_flowY = (flowY * ventral_factor) / dt;
+
+  //TODO: deal with (unlikely) fast changes in Flow?
+
+  // low-pass filter the divergence:
+  of_hover.flowX += (new_flowX - of_hover.flowX) * lp_factor;
+  of_hover.flowY += (new_flowY - of_hover.flowY) * lp_factor;
+
+
+  /***********
+   * CONTROL
+   ***********/
+
+  if (!oscillatingX) {
+    // if not oscillating, increase gain
+    of_hover_ctrl_X.PID.P += of_hover_ctrl_X.ramp * dt;
+  }
+  if (!oscillatingY) {
+    // if not oscillating, increase gain
+    of_hover_ctrl_Y.PID.P += of_hover_ctrl_Y.ramp * dt;
+  }
+
+  // set desired pitch en roll
+  if (oscphi) {
+    of_hover_ctrl_X.errors.err = of_hover_ctrl_X.setpoint - of_hover.flowX;
+    des_inputs.phi = of_hover_ctrl_X.nominal_value + PID_flow_control(dt, &of_hover_ctrl_X);
+  }
+  if (osctheta) {
+    of_hover_ctrl_Y.errors.err = of_hover_ctrl_Y.setpoint - of_hover.flowY;
+    des_inputs.theta = of_hover_ctrl_Y.nominal_value + PID_flow_control(dt, &of_hover_ctrl_Y);
+  }
+
+  // update covariance
+  set_cov_flow(cov_method, &historyX, &historyY, &des_inputs, &covariances);
+  ofh_sp_eu.phi = BFP_OF_REAL(RadOfDeg(des_inputs.phi * oscphi), INT32_ANGLE_FRAC);
+  ofh_sp_eu.theta = BFP_OF_REAL(RadOfDeg(des_inputs.theta * osctheta), INT32_ANGLE_FRAC);
+
+  // Check for oscillations
+  if ((covariances.X < of_hover_ctrl_X.cov_setpoint) && (!oscillatingX)) {
+    oscillatingX = 1;
+
+    if (hover_method == 0) {
+      //Start the Y axis
+      oscillatingY = 0;
+      of_hover_ctrl_Y.PID.P = OFH_PGAINY;
+    }
+    of_hover_ctrl_X.PID.P = of_hover_ctrl_X.PID.P * of_hover_ctrl_X.reduction_factor;
+
+    if (XY_SYMMETRICAL) {
+      // also set Y
+      oscillatingY = 1;
+      of_hover_ctrl_Y.PID.P = of_hover_ctrl_X.PID.P;
+    }
+  }
+  if ((covariances.Y < of_hover_ctrl_Y.cov_setpoint) && (!oscillatingY)) {
+    oscillatingY = 1;
+    of_hover_ctrl_Y.PID.P = of_hover_ctrl_Y.PID.P * of_hover_ctrl_Y.reduction_factor;
+  }
+
+  // Compute 0, 1 or 2 horizontal axes with optitrack
+  computeOptiTrack(!oscphi, !osctheta, &ofh_sp_eu);
```

Comment:
This is because the quad is calibrated by fitting the instability gains vs height slopes, this requires an accurate height measurement, while keeping the other axes stable.